Data/AuthenticationClient/Tests/AuthenticationClientTests.swift

```diff
@@ -0,0 +1,302 @@
+
+        let clientIDHeader = result.allHTTPHeaderFields?.first { $0.key.contains("client-id") }
+        XCTAssertNotNil(clientIDHeader, "Expected to return the client id")
+        XCTAssertTrue(clientIDHeader?.value.contains(configuration.clientID) ?? false, "Expeccted to use the client id")
+    }
+
+    func testAuthenticatingRequestFailsGettingToken() async throws {
+        let state = AutehenticationDataMock()
+        state.accessToken = "abcd"
+        persistence.data = try encoder.encode(state)
+
+        oauthService.refreshResult = .success(nil)
+        _ = try await sut.restoreState()
+
+        let inputRequest = URLRequest(url: URL(string: "https://example.com")!)
+
+        oauthService.accessTokenBehavior = .failure(OAuthServiceError.failedToRefreshTokens(nil))
+
+        try await AsyncAssertThrows(
+            await { _ = try await sut.authenticate(request: inputRequest) }(),
+            nil,
+            "Expected to throw an error as well."
+        )
+        await AsyncAssertEqual(
+            await sut.authenticationState,
+            .notAuthenticated,
+            "Expected to signal not authenticated state"
+        )
+    }
+
+    func testRefreshedTokens() async throws {
+        let state = AutehenticationDataMock()
+        state.accessToken = "abcd"
+        persistence.data = try encoder.encode(state)
+        let newState = AutehenticationDataMock()
+        newState.accessToken = "xyz"
+        oauthService.refreshResult = .success(newState)
+        oauthService.accessTokenBehavior = .success("xyz")
+        _ = try await sut.restoreState()
+
+        let decoded = try persistence.data.map(encoder.decode) as? AutehenticationDataMock
+        XCTAssertEqual(
+            decoded?.accessToken,
+            "xyz",
+            "Expected to persist the refreshed state"
+        )
+
+        let inputRequest = URLRequest(url: URL(string: "https://example.com")!)
+        let resultRequest = try await sut.authenticate(request: inputRequest)
+        let authHeader = resultRequest.allHTTPHeaderFields?.first { $0.key.contains("auth-token") }
+        XCTAssertEqual(authHeader?.value, "xyz", "Expected to use the refreshed access token for the request")
+    }
+
+    // MARK: Private
+
+    private var sut: AuthenticationClientImpl!
+    private var oauthService: OAuthServiceMock!
+    private var encoder: OAuthStateDataEncoder!
+    private var persistence: PersistenceMock!
+}
+
+private struct OauthStateEncoderMock: OAuthStateDataEncoder {
+    func encode(_ data: any OAuthStateData) throws -> Data {
+        guard let data = data as? AutehenticationDataMock else {
+            fatalError()
+        }
+        return try JSONEncoder().encode(data)
+    }
+
+    func decode(_ data: Data) throws -> any OAuthStateData {
+        try JSONDecoder().decode(AutehenticationDataMock.self, from: data)
+    }
+}
+
+private final class OAuthServiceMock: OAuthService {
```

Comment:
Let's move this to Core because we will need to use it in more tests as we integrate auth with other features.